Add functorch M1 shard

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #85565
* #85564

functorch should have a test wherever regular PyTorch gets tested. This
PR adds an M1 shard to test functorch.

Test Plan:
- wait for CI

zou3519 committed:
Add functorch M1 shard

functorch should have a test wherever regular PyTorch gets tested. This
PR adds an M1 shard to test functorch.

Test Plan:
- wait for CI

[ghstack-poisoned]

zou3519 commented:
not ready for review yet

zou3519 committed:
Update on "Add functorch M1 shard"


functorch should have a test wherever regular PyTorch gets tested. This
PR adds an M1 shard to test functorch.

Test Plan:
- wait for CI

[ghstack-poisoned]

zou3519 committed:
Update on "Add functorch M1 shard"


functorch should have a test wherever regular PyTorch gets tested. This
PR adds an M1 shard to test functorch.

Test Plan:
- wait for CI

[ghstack-poisoned]

zou3519 committed:
Update on "Add functorch M1 shard"


functorch should have a test wherever regular PyTorch gets tested. This
PR adds an M1 shard to test functorch.

Test Plan:
- wait for CI

[ghstack-poisoned]

zou3519 commented on .github/workflows/_mac-test.yml:
```diff
@@ -77,6 +77,7 @@ jobs:
       PR_BODY: ${{ github.event.pull_request.body }}
       PYTORCH_RETRY_TEST_CASES: 1
       PYTORCH_OVERRIDE_FLAKY_SIGNAL: 1
+      CONTINUE_THROUGH_ERROR: true
```
I'll delete this before merging...

zou3519 commented:
Should be ready for review

zou3519 commented:
Some notes for the reviewers:
- the functorch M1 shard takes 1 hour. Not sure if this impacts TTS.
- artifacts: AFICT, none of the macos jobs, including the one added from this PR, have artifacts (https://hud.pytorch.org/pr/85565), though I might not be looking in the right place

malfet reviewed (approved):
LGTM, though it does not seem to be adding the shard, is it?

zou3519 commented:
> LGTM, though it does not seem to be adding the shard, is it?

Oops, I accidentally reverted that change. Will put it back, thanks for pointing it out.

zou3519 committed:
Update on "Add functorch M1 shard"


functorch should have a test wherever regular PyTorch gets tested. This
PR adds an M1 shard to test functorch.

Test Plan:
- wait for CI

[ghstack-poisoned]

zou3519 committed:
Update on "Add functorch M1 shard"


functorch should have a test wherever regular PyTorch gets tested. This
PR adds an M1 shard to test functorch.

Test Plan:
- wait for CI

[ghstack-poisoned]

huydhn reviewed (approved):
LGTM!

huydhn commented:
>     * the functorch M1 shard takes 1 hour. Not sure if this impacts TTS.

Looking at the job https://github.com/pytorch/pytorch/actions/runs/3131300487/jobs/5082787746, it looks like pytest-shard works well for this as you can see 3 processes running in parallel there. That was done by https://github.com/pytorch/pytorch/pull/85321 and it was 2+ hours before.  1 hour is good :)

>     * artifacts: AFICT, none of the macos jobs, including the one added from this PR, have artifacts (https://hud.pytorch.org/pr/85565), though I might not be looking in the right place

This is a known connectivity issue between M1 runner and S3 IIRC cc @seemethere

huydhn commented:
On the other hand, you can still get the artifacts from GItHub PR itself https://github.com/pytorch/pytorch/actions/runs/3131300487 (you need to scroll down or just click on the artifacts number there)

![Screen Shot 2022-09-27 at 14 34 09](https://user-images.githubusercontent.com/475357/192639974-ead5aea8-d733-4fcb-b533-11d2ba900f11.png)

zou3519 commented:
Thanks for pointing out the parallelism and the artifacts, @huydhn

zou3519 commented:
@pytorchbot merge
